CRM/Financial/BAO/Payment.php

```diff
@@ -56,62 +56,53 @@ public static function create($params) {
-      $trxn = CRM_Core_BAO_FinancialTrxn::getPartialPaymentTrxn($contribution, $params);
-      if (CRM_Utils_Array::value('line_item', $params) && !empty($trxn)) {
-        foreach ($params['line_item'] as $values) {
-          foreach ($values as $id => $amount) {
-            $p = ['id' => $id];
-            $check = CRM_Price_BAO_LineItem::retrieve($p, $defaults);
-            if (empty($check)) {
-              throw new API_Exception('Please specify a valid Line Item.');
-            }
-            // get financial item
-            $sql = "SELECT fi.id
-            FROM civicrm_financial_item fi
-            INNER JOIN civicrm_line_item li ON li.id = fi.entity_id and fi.entity_table = 'civicrm_line_item'
-            WHERE li.contribution_id = %1 AND li.id = %2";
-            $sqlParams = [
-              1 => [$params['contribution_id'], 'Integer'],
-              2 => [$id, 'Integer'],
-            ];
-            $fid = CRM_Core_DAO::singleValueQuery($sql, $sqlParams);
-            // Record Entity Financial Trxn
-            $eftParams = [
-              'entity_table' => 'civicrm_financial_item',
-              'financial_trxn_id' => $trxn->id,
-              'amount' => $amount,
-              'entity_id' => $fid,
-            ];
-            civicrm_api3('EntityFinancialTrxn', 'create', $eftParams);
+
+    $trxn = CRM_Contribute_BAO_Contribution::recordPartialPayment($contribution, $params);
+    if (CRM_Utils_Array::value('line_item', $params) && !empty($trxn)) {
+      foreach ($params['line_item'] as $values) {
+        foreach ($values as $id => $amount) {
+          $p = ['id' => $id];
+          $check = CRM_Price_BAO_LineItem::retrieve($p, $defaults);
+          if (empty($check)) {
+            throw new API_Exception('Please specify a valid Line Item.');
           }
+          // get financial item
+          $sql = "SELECT fi.id
+          FROM civicrm_financial_item fi
+          INNER JOIN civicrm_line_item li ON li.id = fi.entity_id and fi.entity_table = 'civicrm_line_item'
+          WHERE li.contribution_id = %1 AND li.id = %2";
+          $sqlParams = [
+            1 => [$params['contribution_id'], 'Integer'],
+            2 => [$id, 'Integer'],
+          ];
+          $fid = CRM_Core_DAO::singleValueQuery($sql, $sqlParams);
+          // Record Entity Financial Trxn
+          $eftParams = [
+            'entity_table' => 'civicrm_financial_item',
+            'financial_trxn_id' => $trxn->id,
+            'amount' => $amount,
+            'entity_id' => $fid,
+          ];
+          civicrm_api3('EntityFinancialTrxn', 'create', $eftParams);
         }
       }
-      elseif (!empty($trxn)) {
-        CRM_Contribute_BAO_Contribution::assignProportionalLineItems($params, $trxn->id, $contribution['total_amount']);
-      }
+    }
+    elseif (!empty($trxn)) {
+      CRM_Contribute_BAO_Contribution::assignProportionalLineItems($params, $trxn->id, $contribution['total_amount']);
+    }
```

Comment:
@eileenmcnaughton we need to calculate the check here as per the previous workflow. It should be:
```php
    $paid = CRM_Core_BAO_FinancialTrxn::getTotalPayments($params['contribution_id']);
    $cmp = bccomp($contribution['total_amount'], $paid, 5);
    if ($cmp == 0 || $cmp == -1) {
     civicrm_api3('Contribution', 'completetransaction', ['id' => $contribution['id']]);
    }
```
This would fix the test failures.
NOTE: here the $params['total_amount'] is financial_trxn.total_amount and we shoulddn't compare directly with contribution's total_amount, instead we should compare with total paid.